components/widgets/users/avatars/avatars.tsx

```diff
@@ -11,34 +20,88 @@ import SimpleTooltip, {useSynchronizedImmediate} from 'components/widgets/simple
 import Avatar from 'components/widgets/users/avatar';
 
 import './avatars.scss';
+import {imageURLForUser} from 'utils/utils';
+import {RequireOnlyOne} from 'utils/conditional_types';
 
 type StylingKeys = 'size';
 
 type UserProps = ComponentProps<typeof Avatar> & {name: string};
+type AvatarProps = Omit<UserProps, StylingKeys>;
+type Participants = Array<{id: $ID<UserProfile>}>;
 
-export type Props = Pick<UserProps, StylingKeys> & {
-    users: Array<Omit<UserProps, StylingKeys>>;
-    totalUsers?: number;
-    breakAt?: number;
-}
+export type Props =
+    Pick<UserProps, StylingKeys> &
+    RequireOnlyOne<{
+        participants: Participants;
+        users: AvatarProps[];
+    }> & {
+        totalUsers?: number;
+        breakAt?: number;
+    }
 
 const OTHERS_DISPLAY_LIMIT = 99;
 
+function selectUsers(
+    state: GlobalState,
+    users: Props['users'],
+    participants: Props['participants'],
+): null | {users: AvatarProps[]; missingProfiles?: string[]} {
+    if (users) {
+        return {users};
+    }
+    if (!participants) {
+        return null;
+    }
+
+    const getDisplayName = makeGetDisplayName();
+
+    return participants.reduce<NonNullable<ReturnType<typeof selectUsers>>>((result, {id}) => {
```

Comment:
This returns a new object every time, which means this selector forces a React reconciliation every time, which may be a performance issue (i.e. "slow typing issue" recently worked around). But maybe that's what we're trying to avoid above with the `users` check?

Should we move all this into the render method instead, and only select the minimum necessary data from the store and do everything consistently at render time?

Reply:
The first implementation of this actually did just map in the render method, but there it isn't able to take username display preference into account. 

I knew there was a perf issue with returning an obj/arr here, but doing the selection/mapping at this level _seemed_ best at the time.

Last night I was brainstorming on another redo of this—I think I'll:
- hoist the missing profiles logic and use `getMissingProfilesByIds` higher up (API layer)
- lower/defer `userId`→`Avatar` props selection into `Avatar` and add new selector for `overflowUsers`→namesOverflow tooltip